server/src/main/java/org/jboss/as/server/moduleservice/ExternalModuleSpecService.java

```diff
@@ -32,42 +46,80 @@
         this.moduleIdentifier = moduleIdentifier;
         this.file = file;
+        this.jarFiles = new ArrayList<>();
     }
 
     @Override
     public synchronized void start(StartContext context) throws StartException {
+        final ModuleSpec.Builder specBuilder;
+        String currentName = "";
         try {
-            this.jarFile = new JarFile(file);
+            if (!file.isDirectory()) {
+                currentName = file.toString();
+                this.jarFiles.add(new JarFile(file));
+                specBuilder = ModuleSpec.build(moduleIdentifier.toString());
+                addResourceRoot(specBuilder, jarFiles.get(0));
+                log.debugf("Added %s jar file as resource root for %s module identifier", file.getAbsolutePath(), moduleIdentifier.getName());
+            } else {
+                specBuilder = ModuleSpec.build(moduleIdentifier.toString());
+                final Path rootPath = file.toPath();
+
+                //This path resource root added here pointing to the rootPath will be able to find other file resources in any rootPath subdirectories.
+                //We do not need to add a addPathResourceRoot for each subdirectory
+                addPathResourceRoot(specBuilder, rootPath);
+                log.debugf("Added %s directory as resource root for %s module identifier", file.getAbsolutePath(), moduleIdentifier.getName());
+
+                //scan any jar in this directory and subdirectories
+                final List<Path> processedPaths = new ArrayList<>();
+                try (Stream<Path> pathStream = Files.walk(rootPath).filter(path -> !Files.isDirectory(path) && path.getFileName().toString().endsWith(".jar"))) {
+                    pathStream.forEach(p -> {
+                        processedPaths.add(p);
+                        if (processedPaths.size() > MAX_NUMBER_OF_JAR_RESOURCES)
+                            throw new RuntimeException(ServerLogger.ROOT_LOGGER.maximumNumberOfJarResources(specBuilder.getName(), MAX_NUMBER_OF_JAR_RESOURCES));
+                    });
+                }
+
+                if (! processedPaths.isEmpty()) {
+                    final TreeSet<Path> jars = new TreeSet<>(Comparator.comparing(Path::getFileName).thenComparing(this::fileSize));
+                    jars.addAll(processedPaths);
+                    for (Path jar : jars) {
+                        currentName = jar.toString();
+                        JarFile jarFile = new JarFile(jar.toFile());
+                        this.jarFiles.add(jarFile);
+                        addResourceRoot(specBuilder, jarFile);
+                        log.debugf("Added %s jar file as resource root for %s module identifier", jar.toString(), moduleIdentifier.getName());
+                    }
+                }
+            }
         } catch (IOException e) {
+            throw ServerLogger.ROOT_LOGGER.errorOpeningZipFile(currentName);
```

Comment:
Shouldn't this pass in the underlying IOException so it can be logged?

Reply:
Correct, adding the exception to the error trace will log the original exception stacktrace too